src/libbson/src/bson/bson-utf8.c

```diff
@@ -264,77 +376,104 @@ bson_utf8_escape_for_json (const char *utf8, /* IN */
    bson_unichar_t c;
    bson_string_t *str;
    bool length_provided = true;
+   ssize_t pos;
    const char *end;
 
    BSON_ASSERT (utf8);
 
-   str = bson_string_new (NULL);
-
    if (utf8_len < 0) {
       length_provided = false;
       utf8_len = strlen (utf8);
    }
 
+   if (utf8_len == 0) {
+      return bson_strdup ("");
+   }
+
    end = utf8 + utf8_len;
 
-   while (utf8 < end) {
+   str = bson_string_alloc (utf8_len);
+
+   pos = 0;
+
+   do {
       uint8_t mask;
       uint8_t length_of_char;
 
+      c = (unsigned char) utf8[pos];
+      if (!_is_special_char (c)) {
+         // Normal character, no need to do anything besides iterate
+         // Copy rest of the string if we reach the end
+         pos++;
+         utf8_len--;
+         if (utf8_len == 0) {
+            bson_string_append_ex (str, utf8, pos);
+            break;
+         }
+
+         continue;
+      }
+
+      // Reached a special character. Copy over all of normal characters
+      // we have passed so far
+      if (pos) {
+         bson_string_append_ex (str, utf8, pos);
+         utf8 += pos;
+         pos = 0;
+      }
+
       // Check if expected char length goes past end
+      // bson_utf8_get_char will crash without this check
       _bson_utf8_get_sequence (utf8, &length_of_char, &mask);
       if (utf8 > end - length_of_char) {
          goto invalid_utf8;
       }
 
-      c = bson_utf8_get_char (utf8);
+      // Check for null character
+      // Null characters are only allowed if the length is provided
+      if (!*utf8 || (*utf8 == '\xc0' && *(utf8 + 1) == '\x80')) {
+         if (!length_provided) {
+            goto invalid_utf8;
+         }
 
-      switch (c) {
-      case '\\':
-      case '"':
-         bson_string_append_c (str, '\\');
-         bson_string_append_unichar (str, c);
-         break;
-      case '\b':
-         bson_string_append (str, "\\b");
-         break;
-      case '\f':
-         bson_string_append (str, "\\f");
-         break;
-      case '\n':
-         bson_string_append (str, "\\n");
-         break;
-      case '\r':
-         bson_string_append (str, "\\r");
-         break;
-      case '\t':
-         bson_string_append (str, "\\t");
-         break;
-      default:
-         if (c < ' ') {
-            bson_string_append_printf (str, "\\u%04x", (unsigned) c);
-         } else {
-            bson_string_append_unichar (str, c);
+         bson_string_append (str, "\\u0000");
+         utf8_len -= *utf8 ? 2 : 1;
+         utf8 += *utf8 ? 2 : 1;
+         continue;
+      }
+
+      // Unicode character
+      if (c > 0x7f /* highest ASCII character */) {
```

Comment:
```suggestion
      // Highest ASCII character.
      if (c > 0x7fu) {
```
Signedness consistency + should the block scope comment be updated?